### PR TITLE
Shebang flakes v2

### PR DESCRIPTION
**Pull request by roberth**
# Motivation

This adds shebang support as implemented by @tomberek with the changes/additions
 - Double backtick quoting only. This keeps things simple, while also delivering a powerful syntax for self contained scripts (one file) with custom expressions.
 - Relative paths are relative to the script. Otherwise, they'd be relative to the working directory, which is would be unpredictable and therefore less useful.
 - It reserves syntax in the shebang lines; in other words, some symbols demand quotes. This has two benefits
    - The syntax looks unambiguous. Quoted symbols look like literal symbols, and they are.
    - The syntax can be extended later, without breaking scripts that used to work.

Remaining todo:
 - [x] last commit makes flake flags relative to script dir. tests?
 - [ ] should some commands add an implicit `--`? All commands or just `run`/`shell`/`develop`?

# Context

Based on
- #5189 

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [x] agreed on idea
 - [x] agreed on implementation strategy
 - [x] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [x] documentation in the manual
 - [x] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [x] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

File: src/nix/shell.md [hunk 1]
```diff
@@ -51,4 +51,120 @@ R""(
...
+for n in range(1, 10): t.add_row([n, n * n])
+print t
+```
+
+Similarly, the following is a Perl script that specifies that it
+requires Perl and the `HTML::TokeParser::Simple` and `LWP` packages:
+
+```perl
+#! /usr/bin/env nix
+#! nix shell github:tomberek/-#perlWith.HTMLTokeParserSimple.LWP --command perl -x
```
Comment (fricklerhandwerk) [inline src/nix/shell.md:94]: What does the `-x` belong to? It looks like it's passed to `perl`, but how would Nix know?

**Review by fricklerhandwerk — COMMENTED** [review]
From a superficial perspective of a potential user and manual reader, I mainly have two questions:
1. Why is that "weird" escaping needed? If it's technically unavoidable it sure needs the reason documented. Otherwise it appears arbitrary and cumbersome.
2. Why is it `nix shell`? What does that have to do with shells at all?

I understand that historically stuff was just tacked onto `nix-shell` being a kitchen sink, but this is a different situation. Could we do something like `#! nix shebang ... --interpreter`?

**Review by tomberek — CHANGES_REQUESTED** [review]
Remaining items are release notes and docs updates from @fricklerhandwerk

File: src/nix/shell.md [hunk 2]
```diff
@@ -51,4 +51,120 @@ R""(
...
+
+terraform "$@"
+```
+
+> **Note**
+>
+> You must use double backticks (```` `` ````) when passing a simple Nix expression
+> in a nix shell shebang.
+
+Finally, using the merging of multiple nix shell shebangs the following
```
[the remark this answers is not in the document]
Comment (tomberek) [reply src/nix/shell.md:126]: multiple lines of "#!nix" are concatenated after that prefix is removed. This allows breaking up a long line into multiple lines.

File: src/nix/shell.md [hunk 3]
```diff
@@ -51,4 +51,120 @@ R""(
...
+import Text.HTML.TagSoup
+import Data.Either
+import Data.ByteString.Char8 (unpack)
+
+-- Fetch nixos.org and print all hrefs.
+main = do
+  resp <- openURI "https://nixos.org/"
+  let tags = filter (isTagOpenName "a") $ parseTags $ unpack $ fromRight undefined resp
+  let tags' = map (fromAttrib "href") tags
+  mapM_ putStrLn $ filter (/= "") tags'
```
[the remark this answers is not in the document]
Comment (tomberek) [reply src/nix/shell.md:145]: history: I tried to copy the previous examples and modify as little as possible. I am not opposed to a simpler script if that would communicate the concept better.

File: src/nix/shell.md [hunk 4]
```diff
@@ -51,4 +51,120 @@ R""(
...
+#! nix ``
+#! nix --command bash
+
+terraform "$@"
+```
+
+> **Note**
+>
+> You must use double backticks (```` `` ````) when passing a simple Nix expression
+> in a nix shell shebang.
```
[the remark this answers is not in the document]
Comment (iFreilicht) [reply src/nix/shell.md:124]: I guess the idea is that a regular `nix shell` invocation can easily be translated into a shebang?

If you start by experimenting on the CLI and end up with a line like

```
nix shell nixpkgs#terraform nixpkgs#nodejs_20 --command python myscript.py
```

You can move that into a shebang without changing much.

File: doc/manual/src/release-notes/rl-next.md [hunk 5]
```diff
@@ -1,2 +1,52 @@
...
+* The experimental nix command is now a `#!-interpreter` by appending the
+  contents of any `#! nix` lines and the script's location to a single call.
+
+  Verbatim strings may be passed in double backtick (```` `` ````) quotes.
+  `--file` and `--expr` resolve relative paths based on the script location.
+
+  Some examples:
+  ```
+  #!/usr/bin/env nix
+  #! nix shell --file ``<nixpkgs>`` hello --command bash
```
Comment (tomberek) [inline doc/manual/src/release-notes/rl-next.md:12]:
```suggestion
  #! nix shell --file ``<nixpkgs>`` hello cowsay --command bash
```

File: src/nix/shell.md (hunk 1 above)
Comment (tomberek) [reply src/nix/shell.md:94]: this is part of the "--command"

**Comment by tomberek** [comment]
I started trying to incorporate the suggestions. (https://github.com/hercules-ci/nix/pull/1/commits/5d33afe98751f861bcc2a61ef2d1de18f4e59f81)

File: tests/flakes/flakes.sh [hunk 6]
```diff
@@ -64,9 +64,82 @@ cat > $nonFlakeDir/README.md <<EOF
...
 
+# this also tests a fairly trivial double backtick quoted string, ``--command``
+cat > $nonFlakeDir/shebang-comments.sh <<EOF
+#! $(type -P env) nix
+# some comments
+# some comments
+# some comments
+#! nix --offline shell
+#! nix flake1#fooScript
+#! nix --no-write-lock-file ``--command`` bash
```
Comment (tomberek) [inline tests/flakes/flakes.sh:89]: superfluous?

File: src/libutil/args.cc [hunk 7]
```diff
@@ -74,7 +77,148 @@ std::optional<std::string> RootArgs::needsCompletion(std::string_view s)
...
+            return;
+        }
+        switch (remaining[0]) {
+            case ' ': case '\t': case '\n': case '\r':
+                if (!acc.empty())
+                    r.push_back(acc);
+                state = std::make_shared<ParseUnquoted>(ParseUnquoted(remaining.substr(1)));
+                return;
+            case '`':
+                if (remaining.size() > 1 && remaining[1] == '`') {
```
Comment (tomberek) [inline src/libutil/args.cc:138]: Can check to see if there is a third "`" or not?

Comment (roberth) [reply src/libutil/args.cc:141]: Done. It now has a property test to make sure all strings are representable.

Triple quotes now need to be written as four quotes.

``````nix
#! nix --argstr markdown ``
#! nix ````nix
#! nix # identity
#! nix a: a
#! nix ````
``````

**Review by tomberek — APPROVED** [review]

**Comment by Ericson2314** [comment]
FYI perhaps we never made this policy, but this was merged (a) without a merge commit and (b) with commits that perhaps don't individually build. The combination of those two things makes bisecting later harder. (Merge commit helps with more complicated history because `--first-parent` allows skipping the PR history.)

**Comment by tomberek** [comment]
> FYI perhaps we never made this policy, but this was merged (a) without a merge commit and (b) with commits that perhaps don't individually pass tests (my guess from the messages). The combination of those two things makes bisecting later harder. (Merge commit helps with more complicated history because `--first-parent` allows skipping the PR history.)

Worth a revert and re-merge?

**Comment by roberth** [comment]
Three PRs have been merged since, and would have to be remerged, plus you get the general nastiness of force-pushing to master. I'd say leave as is, but pay attention to how you're merging PRs.
Assuming you're using the green button, that thing is stateful. It remembers which way you merged the last one, so just always leave it at "merge" and never micro optimize by using rebase for single commit PRs.
Squashing should be done with the CLI, preferably by the author.

**Pull request merged**